core/src/main/java/org/apache/lucene/search/grouping/CollapsingDocValuesSource.java

```diff
@@ -0,0 +1,215 @@
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+
+import java.io.IOException;
+
+/**
+ * Utility class that ensures that a single collapse key is extracted per document.
+ */
+abstract class CollapsingDocValuesSource<T> {
+    protected final String field;
+
+    CollapsingDocValuesSource(String field) throws IOException {
+        this.field = field;
+    }
+
+    abstract T get(int doc);
+
+    abstract T copy(T value, T reuse);
+
+    abstract void setNextReader(LeafReader reader) throws IOException;
+
+    /**
+     * Implementation for {@link NumericDocValues} and {@link SortedNumericDocValues}.
+     * Fails with an {@link IllegalStateException} if a document contains multiple values for the specified field.
+     */
+    static class Numeric extends CollapsingDocValuesSource<Long> {
+        private NumericDocValues values;
+        private Bits docsWithField;
+
+        public Numeric(String field) throws IOException {
+            super(field);
+        }
+
+        @Override
+        public Long get(int doc) {
+            if (docsWithField.get(doc)) {
+                return values.get(doc);
+            } else {
+                return null;
+            }
+        }
+
+        @Override
+        public Long copy(Long value, Long reuse) {
+            if (value == null) {
+                return null;
+            }
+            return value.longValue();
+        }
+
+        @Override
+        public void setNextReader(LeafReader reader) throws IOException {
+            DocValuesType type = getDocValuesType(reader, field);
+            if (type == null) {
+                values = DocValues.emptyNumeric();
+                docsWithField = new Bits.MatchNoBits(reader.maxDoc());
+                return ;
+            }
+            docsWithField = DocValues.getDocsWithField(reader, field);
+            switch (type) {
+                case NUMERIC:
+                    values = DocValues.getNumeric(reader, field);
+                    break;
+
+                case SORTED_NUMERIC:
+                    final SortedNumericDocValues sorted = DocValues.getSortedNumeric(reader, field);
+                    values = DocValues.unwrapSingleton(sorted);
+                    if (values == null) {
+                        values = new NumericDocValues() {
+                            @Override
+                            public long get(int docID) {
+                                sorted.setDocument(docID);
+                                assert sorted.count() > 0;
+                                if (sorted.count() > 1) {
+                                    throw new IllegalStateException("failed to collapse " + docID +
+                                        ", the collapse field must be single valued");
+                                }
+                                return sorted.valueAt(0);
```

Comment:
Can we return 0 when the value count is 0 to be consistent with the singe-value case, or throw a proper exception? I am concerned this code could raise a weird error message otherwise.